packages/smart-contracts/src/contracts/ERC20TransferrableReceivable.sol

```diff
@@ -0,0 +1,170 @@
+// Contract based on https://docs.openzeppelin.com/contracts/3.x/erc721
+// SPDX-License-Identifier: MIT
+pragma solidity ^0.8.0;
+
+import '@openzeppelin/contracts/utils/Counters.sol';
+import '@openzeppelin/contracts/token/ERC721/extensions/ERC721URIStorage.sol';
+
+contract ERC20TransferrableReceivable is ERC721URIStorage {
+  using Counters for Counters.Counter;
+
+  // Counter for uniquely identifying payments
+  Counters.Counter private _paymentId;
+
+  // Counter for uniquely identifying receivable tokens
+  Counters.Counter private _receivableTokenId;
+
+  struct ReceivableInfo {
+    address tokenAddress;
+    uint256 amount;
+    uint256 balance;
+  }
+  mapping(uint256 => ReceivableInfo) public receivableInfoMapping;
+
+  // Nested mapping for looking up receivable token given a paymentReference
+  // and original payment address
+  mapping(bytes32 => uint256) public receivableTokenIdMapping;
+
+  // Helper mappings to lookup all receivable tokens owned by an address
+  mapping(address => uint256[]) private _receivableIds;
+  mapping(uint256 => uint256) private _receivableIdIndexes;
+
+  address public paymentProxy;
+
+  // Event to declare payments to a receivableTokenId
+  event Payment(
+    address sender,
+    address recipient,
+    address paymentProxy,
+    uint256 receivableTokenId,
+    address tokenAddress,
+    uint256 paymentId,
+    bytes indexed paymentReference
+  );
+
+  // Event to declare a transfer with a reference
+  // This event is emitted from a delegatecall to an ERC20FeeProxy contract
+  event TransferWithReferenceAndFee(
+    address tokenAddress,
+    address to,
+    uint256 amount,
+    bytes indexed paymentReference,
+    uint256 feeAmount,
+    address feeAddress
+  );
+
+  constructor(
+    string memory name,
+    string memory symbol,
+    address _paymentProxyAddress
+  ) ERC721(name, symbol) {
+    paymentProxy = _paymentProxyAddress;
+  }
+
+  function supportsInterface(bytes4 interfaceId) public view override(ERC721) returns (bool) {
+    return super.supportsInterface(interfaceId);
+  }
+
+  function payOwner(
+    uint256 receivableTokenId,
+    uint256 amount,
+    bytes calldata paymentReference,
+    uint256 feeAmount,
+    address feeAddress
+  ) external {
+    require(amount != 0, 'Zero amount provided');
+    address owner = ownerOf(receivableTokenId);
+    _paymentId.increment();
+
+    ReceivableInfo storage receivableInfo = receivableInfoMapping[receivableTokenId];
+    address tokenAddress = receivableInfo.tokenAddress;
+
+    (bool status, ) = paymentProxy.delegatecall(
+      abi.encodeWithSignature(
+        'transferFromWithReferenceAndFee(address,address,uint256,bytes,uint256,address)',
+        tokenAddress,
+        owner,
+        amount,
+        paymentReference,
+        feeAmount,
+        feeAddress
+      )
+    );
+    require(status, 'transferFromWithReferenceAndFee failed');
+
+    receivableInfo.balance += amount;
```

Comment:
Please make all state changes before external calls ("checks-effects-interactions" pattern). Else, the function may be susceptible to reentrancy attacks.